omniscidb/DataMgr/ChunkMetadata.h

```diff
@@ -154,76 +154,134 @@ inline void mergeStats(ChunkStats& lhs,
+
+    return res;
   }
 
   std::string toString() const {
     return dump();
   }
 #endif
 
-  ChunkMetadata(const hdk::ir::Type* type_,
-                const size_t num_bytes,
-                const size_t num_elements,
-                const ChunkStats& chunk_stats)
-      : type(type_)
-      , numBytes(num_bytes)
-      , numElements(num_elements)
-      , chunkStats(chunk_stats) {}
-
-  ChunkMetadata() {}
-
   template <typename T>
   void fillChunkStats(const T min, const T max, const bool has_nulls) {
-    ::fillChunkStats(chunkStats, type, min, max, has_nulls);
+    StatsMaterializeFn().swap(stats_materialize_fn_);
+    ::fillChunkStats(chunk_stats_, type_, min, max, has_nulls);
+  }
+
+  void fillChunkStats(const ChunkStats& new_stats) {
+    StatsMaterializeFn().swap(stats_materialize_fn_);
+    chunk_stats_ = new_stats;
   }
 
   void fillChunkStats(const Datum min, const Datum max, const bool has_nulls) {
-    chunkStats.has_nulls = has_nulls;
-    chunkStats.min = min;
-    chunkStats.max = max;
+    StatsMaterializeFn().swap(stats_materialize_fn_);
+    chunk_stats_.has_nulls = has_nulls;
+    chunk_stats_.min = min;
+    chunk_stats_.max = max;
+  }
+
+  void fillStringChunkStats(const bool has_nulls) {
+    StatsMaterializeFn().swap(stats_materialize_fn_);
+    chunk_stats_.has_nulls = has_nulls;
+#ifndef __CUDACC__
+    chunk_stats_.min.stringval = nullptr;
+    chunk_stats_.max.stringval = nullptr;
+#endif
   }
 
   bool operator==(const ChunkMetadata& that) const {
-    return type->equal(that.type) && numBytes == that.numBytes &&
-           numElements == that.numElements &&
-           DatumEqual(
-               chunkStats.min,
-               that.chunkStats.min,
-               type->isArray() ? type->as<hdk::ir::ArrayBaseType>()->elemType() : type) &&
-           DatumEqual(
-               chunkStats.max,
-               that.chunkStats.max,
-               type->isArray() ? type->as<hdk::ir::ArrayBaseType>()->elemType() : type) &&
-           chunkStats.has_nulls == that.chunkStats.has_nulls;
+    if (!type_->equal(that.type_) || num_bytes_ != that.num_bytes_ ||
+        num_elements_ != that.num_elements_) {
+      return false;
+    }
+
+    maybeMaterializeStats();
+    that.maybeMaterializeStats();
+
+    return DatumEqual(chunk_stats_.min,
+                      that.chunk_stats_.min,
+                      type_->isArray() ? type_->as<hdk::ir::ArrayBaseType>()->elemType()
+                                       : type_) &&
+           DatumEqual(chunk_stats_.max,
+                      that.chunk_stats_.max,
+                      type_->isArray() ? type_->as<hdk::ir::ArrayBaseType>()->elemType()
+                                       : type_) &&
+           chunk_stats_.has_nulls == that.chunk_stats_.has_nulls;
   }
+
+ private:
+  void maybeMaterializeStats() const {
+    if (stats_materialize_fn_) {
+      stats_materialize_fn_(chunk_stats_);
+      StatsMaterializeFn().swap(stats_materialize_fn_);
+    }
```

Comment:
Should there be some logging or fatal error for failure mode here?

Reply:
I'd expect any logging and exceptions to be in the materializer because that's where synchronization happens. But I don't expect stats computation ever fail because we can always use the default stats.